### DISCUSSION TITLE
Alternate build of Redbean without ssl and sqlite3

### DISCUSSION THREAD

**Discussion by yapcheahshen** [jart/cosmopolitan#472] [Ideas]
It seems like sqlite3 takes up nearly 1MB, and it is not as essential as Lua.

**Comment by jart** [comment]
One of Redbean's values is you should only have to pay for what you use. For that reason, we do exactly what you propose. See https://redbean.dev/ where we have an "alternative build" called redbean-original. That release is only half a meg rather than 2 megs. If you build it in MODE=tiny then it's able to 222kb. If you build it in MODE=tinylinux (non-portable) then it's 170kb. Hope you enjoy! Let me know if there's more feedback.

**Reply by yapcheahshen** [reply]
Redbean-original is really impressive and completely fits my need except the LaunchBrowser feature,
appreciate your great work after years of putting up with Electron , NodeWebkit and many other so-called cross platform toolkits.

As [launchbrowser](https://github.com/jart/cosmopolitan/blob/master/tool/net/redbean.c#L7258) is initialized by Lua, it will be nice if *.args* can specifiy a start up url.
how about add one more line in the switch statement of *redbean.c#GetOpts* ?
`CASE('w', launchbrowser = strdup(optarg));`

**Reply by jart** [reply]
Nice suggestion. I did exactly the thing you proposed in https://github.com/jart/cosmopolitan/commit/5b11033d4dde698ec0b0850a4c9d1bdd8d90726d

**Reply by yapcheahshen** [reply]
readbean-original-2.0.14 works fine , thank you.
I study redbean.c and found that there is no way to list all files in local directory (outside of the zip container) specified by -D flag.
is that because of security concern?
if so, the Lua-only build (no Sqlite, no SSL) of Redbean is still in need, as Lua is the only way to access cosmopolitan APIs (namely , unix.opendir ) .

**Reply by jart** [reply]
Use `unix.opendir`

**Comment by Alamantus** [comment]
Sorry to pop in on this, but how would I go about building a version of redbean.com that includes only Lua and not sqlite or security? I took a peek at the `redbean-original.c` file and some other redbean files in `tools/net`, and all I noticed was defining `STATIC` and `UNSECURE`, but I couldn't see where sqlite was actually being included in `redbean.c` (but then again I'm not great a reading C code).

It looks like `STATIC` disables/excludes Lua and `UNSECURE` disables/excludes SSL stuff, but what about sqlite? `redbean-original` as @jart suggested appears to exclude Lua in addition to sqlite and SSL, so is this particular combination not currently doable / is sqlite intrinsically tied to Lua and can't be excluded separately?

**Reply by yapcheahshen** [reply]
Sqlite is inlcuded as a [Lua third party library](https://github.com/jart/cosmopolitan/blob/master/tool/net/redbean.c#L5298).
we may use a flag to exclude Sqlite.

#ifndef NOSQLITE
{"lsqlite3", luaopen_lsqlite3},   // luaopen_lsqlite3 is the entry of lua-sqlite binding defined in [lsqlite3.c](https://github.com/jart/cosmopolitan/blob/master/tool/net/lsqlite3.c)
#endif

And a new entry main program redbean-nosqlite.c based on[ redbean-unsecure.c](https://github.com/jart/cosmopolitan/blob/master/tool/net/redbean-unsecure.c)

#define UNSECURE  // remove ssl 
#define NOSQLITE  //  remove sqlite 
#define REDBEAN "redbean-nosqlite"  // name of the custom build
#include "tool/net/redbean.c"
